content/testable-architecture.md

```diff
@@ -3,4 +3,67 @@ title: Testable Architecture
 description: Designing for testability
 ---
 
-Testable Architecture
\ No newline at end of file
+Automated tests and executable specifications can only work with software that has a testable architecture.
+A testable architecture enables fast and reliable tests that are easy to write, execute and maintain.
+Fast tests enable the developers to run them frequently to obtain fast feedback on what they are building, without losing focus or flow.
+
+When designing for testability, make sure that your products and services are composed of loosely-coupled,
+well-encapsulated components or modules.
+
+Decoupling your business logic from your infrastructure (the often slow and brittle components),
+makes it possible to test at different levels, with maximum confidence and minimum cost.
+
+* Ensure that you can run your tests without going through the user interface (UI). They are slow, brittle, expensive, and hard to fix. Do not depend solely on UI tests.
+* Running tests against a database makes tests slow. We need to make sure the database is in the expected state before each test in order to make the tests behave consistently.
+This will ensure that tests don't interfere with each other and can be executed in any order.
+
+We want most of our tests to use some kind of in-memory stub implementation, instead of an actual database.
+From the perspective of the domain logic, the behavior looks exactly the same.
+To make sure that the stub works in the same way as the real thing, we need to have confidence, which we can obtain using [contract tests](https://martinfowler.com/bliki/IntegrationContractTest.html).
+
+We can think of our database and our stub as two things that can be plugged into the same socket, or **port**.
```

Comment:
I think the word "socket" can easily be misunderstood as "TCP socket", so maybe clarify that it is a "socket", not a socket :-)

Reply:
Thanks @aslakhellesoy ! Yeah, the "socket" here is used as a metaphor; I'll just remove it to avoid confusion.